File: src/Automaton/Automaton.py
import sys

sys.path.insert(1, "../FLFA-Labs/src/Grammar")
import Grammar


class FiniteAutomaton:
    def __init__(self,
                 states,
                 alphabet,
                 transitions,
                 initial_state,
                 final_sates) -> None:
        self.states = states
        self.alphabet = alphabet
        self.transitions = transitions
        self.initial_state = initial_state
        self.final_states = final_sates

    def RG_to_NFA(self, grammar) -> 'FiniteAutomaton':
        productions = grammar.productions
        states = set(productions.keys()) | {"X"}
        alphabet = set()
        transitions = {}
        initial_state = grammar.initial_symbol()
        final_states = set("X") | \
                       ({initial_state} if "&" in productions[initial_state] else set())
        for non_terminal, prods in productions.items():
            for production in prods:
                if production == "&":
                    continue
                new_transition = "X" if len(production) == 1 else production[1]
                transitions.setdefault(
                    (non_terminal, production[0]), set()
                ).add(new_transition)
                alphabet.add(production[0])
        return FiniteAutomaton(states, alphabet, transitions, initial_state, final_states)

    def check_word(self, word) -> bool:
        current_state = {self.initial_state}
        for symbol in word:
            next_state = set()
            for state in current_state:
                next_state.update(
                    self.transitions.get((state, symbol), set())
                )
            current_state = next_state
        return bool(current_state.intersection(self.final_states))

    def determine_FA(self) -> str:
        transition_count = []
        for k, states in self.transitions.items():
            for state in states:
                if k[1] == '&':
                    return "&-NFA"
                if k[0] in transition_count:
                    return "NFA"
                else:
                    transition_count.append(k[0])
        return "DFA"

    def __closure(self, q) -> list:
        closure = list()
        closure.append(q)
        for element in closure:
            for k, states in self.transitions.items():
                for state in states:
                    if k[0] == element and k[1] == '&':
                        closure.append(state)
        return closure

    def NFA_to_DFA(self):
        # checking if it is already a DFA
        if self.determine_FA() == 'DFA':
            raise ValueError("This is already a DFA")

        # additional functions to avoid nesting
        def add_new_transition():
            nonlocal dfa_state, letter, new_state
            # values in dictionaries are not always traversed in order
            # to avoid duplicates with different order of elements
            # sort new_state before adding it in dfa_states
            new_state.sort()
            if len(new_state) > 0:
                # making a transition if there is a state
                dfa_transitions.setdefault(
                    (tuple(dfa_state), letter), set()
                ).add(tuple(new_state))
                update_final_states()
                # adding the state into dfa_states if it's a new one
                if new_state not in dfa_states:
                    dfa_states.append(new_state)

        def update_final_states():
            nonlocal new_state
            if any(item in new_state for item in self.final_states)\
                    and tuple(new_state) not in dfa_final_states:
                dfa_final_states.append(tuple(new_state))

        def find_new_state():
            nonlocal dfa_state, letter, new_state
            for element in dfa_state:
                for key, states in self.transitions.items():
                    # finding the necessary state in dict key
                    if key[0] == element and letter == key[1]:
                        for state in states:
                            if state not in new_state:
                                new_state = append_closure(state)

        def append_closure(state):
            nonlocal new_state, closures
            for closure in closures:
                if state == closure[0]:
                    new_state = new_state + closure
                    break
            new_state = list(set(new_state))
            return new_state

        # initializing variables of the DFA
        dfa_states = list()
        dfa_transitions = {}
        dfa_final_states = list()
        closures = list()
        # making a list of &-closure
        for key, states in self.transitions.items():
            if key[0] in [item[0] for item in closures]:
                continue
            closures.append(self.__closure(key[0]))
        # dfa_states[0] is the initial state of DFA
        dfa_states.append(closures[0])
        self.initial_state = dfa_states[0]

        for dfa_state in dfa_states:
            for letter in self.alphabet:
                new_state = list()
                find_new_state()
                add_new_transition()

        self.states = dfa_states
        self.transitions = dfa_transitions
        self.final_states = dfa_final_states

File: Go/automaton/automaton.go
package automaton

import (
	"flfa/functions"
	"slices"
	"strings"
)

type Transition struct {
	InitialState string
	Transition   rune
	NextState    string
}

type FiniteAutomaton struct {
	states       []string
	alphabet     []rune
	transitions  []Transition
	initialState string
	finalStates  []string
}

func NewFiniteAutomaton(
	states []string,
	alpahbet []rune,
	transitions []Transition,
	initialState string,
	finalStates []string) *FiniteAutomaton {

	return &FiniteAutomaton{
		states,
		alpahbet,
		transitions,
		initialState,
		finalStates}
}

func (f *FiniteAutomaton) CheckWord(word string) bool {
	var states map[string]bool = map[string]bool{f.initialState: true}
	for _, letter := range word {
		var nextStates map[string]bool = make(map[string]bool)
		for state := range states {
			for _, transition := range f.transitions {
				if state == transition.InitialState && letter == transition.Transition {
					nextStates[transition.NextState] = true
				}
			}
		}
		states = nextStates
	}
	for _, finalState := range f.finalStates {
		if states[finalState] {
			return true
		}
	}
	return false
}

func (f *FiniteAutomaton) DetermineFA() string {
	transitions := make(map[string]rune)
	isNFA := false
	for _, transition := range f.transitions {
		if transition.Transition == '&' {
			return "&-NFA"
		}
		if initialState, ok := transitions[transition.InitialState]; ok &&
			initialState == transition.Transition {
			isNFA = true
		} else {
			transitions[transition.InitialState] = transition.Transition
		}
	}
	if isNFA {
		return "NFA"
	}
	return "DFA"
}

func (f *FiniteAutomaton) closure(state string) map[string]bool {
	closure := map[string]bool{state: true}
	for element := range closure {
		for _, transition := range f.transitions {
			if transition.InitialState == element &&
				transition.Transition == '&' {
				closure[transition.NextState] = true
			}
		}
	}
	return closure
}

func joinStates(states []string) string {
	if len(states) > 1 {
		return "{" + strings.Join(states, ", ") + "}"
	}
	return states[0]
}

func MakeDFA(f FiniteAutomaton) *FiniteAutomaton {
	if f.DetermineFA() == "DFA" {
		return &f
	}
	dfaTransitions := make([]Transition, 0, 10)
	dfaFinalStates := make([]string, 0, 2)
	dfaStatesMap := make([]map[string]bool, 0)

	closures := make(map[string]map[string]bool)
	for _, state := range f.states {
		closures[state] = f.closure(state)
		var newState map[string]bool = closures[state]
		dfaStatesMap = append(dfaStatesMap, newState)
	}

	i := 0
	for i < len(dfaStatesMap) {
		state := dfaStatesMap[i]
		for _, symbol := range f.alphabet {
			newStateMap := make(map[string]bool)
			for element := range state {
				for _, transition := range f.transitions {
					if transition.Transition == symbol &&
						transition.InitialState == element {
						for closure, ok := range closures[transition.NextState] {
							newStateMap[closure] = ok
						}
					}
				}
			}
			if len(newStateMap) > 0 {
				stateSlice := functions.KeyList(state)
				newStateSlice := functions.KeyList(newStateMap)
				dfaTransitions = append(dfaTransitions, Transition{
					joinStates(stateSlice),
					symbol,
					joinStates(newStateSlice),
				})
				if functions.ContainsMap(dfaStatesMap, newStateMap) == false {
					dfaStatesMap = append(dfaStatesMap, newStateMap)
				}
			}
		}
		i++
	}
	dfaStates := make([]string, 0, len(dfaStatesMap))
	for _, stateMap := range dfaStatesMap {
		state := functions.KeyList(stateMap)
		dfaStates = append(dfaStates, joinStates(state))
		for _, finalState := range f.finalStates {
			if slices.Contains(state, finalState) {
				dfaFinalStates = append(dfaFinalStates, joinStates(state))
			}
		}
	}
	return NewFiniteAutomaton(dfaStates, f.alphabet, dfaTransitions, f.initialState, dfaFinalStates)
}
